Code/Tools/FBuild/FBuildCore/Graph/NodeGraph.cpp

```diff
@@ -187,12 +188,35 @@ bool NodeGraph::Load( IOStream & stream, bool & needReparsing )
 	{
 		const AString & fileName = m_UsedFiles[ i ].m_FileName;
 		const uint64_t timeStamp = FileIO::GetFileLastWriteTime( fileName );
-		if ( timeStamp != m_UsedFiles[ i ].m_TimeStamp )
+		if ( timeStamp == m_UsedFiles[ i ].m_TimeStamp )
+			continue; // timestamps match, no need to check hashes
+
+		FileStream fs;
+		if ( fs.Open( fileName.Get(), FileStream::READ_ONLY ) == false )	
 		{
-			FLOG_WARN( "BFF file '%s' has changed (reparsing will occur).", fileName.Get() );
+			FLOG_INFO( "BFF file '%s' missing or unopenable (clean build will result).", fileName.Get() );
 			needReparsing = true;
-			return true;
+			return true; // not opening the file is not an error, it could be not needed anymore
+		}
+
+		const size_t size = (size_t)fs.GetFileSize();
+		AutoPtr< void > mem( ALLOC( size ) );
+		if ( fs.Read( mem.Get(), size ) != size )
+		{
+			return false; // error reading
+		}
+
+		const uint64_t dataHash = xxHash::Calc64( mem.Get(), size );
+		if ( dataHash == m_UsedFiles[ i ].m_DataHash )
+		{
+			// set file modification time to the one saved in the DB to save time on the next run
+			FileIO::SetFileLastWriteTime( fileName, timeStamp );
```

Comment:
Instead of modifying the file stamp, perhaps we should modify m_TimeStamp to match that of the file. This has a couple of benefits:
 - Won't cause dirtying of build for other bff hierarchies using the same bff file
 - Won't impact source control that uses modification time to reflect something (i.e. perforce "modtime" option)